File: ports/go/security/surrogate_reassembly_test.go
package security

import "testing"

// Ground truth: the detect_* spot-check theorems in
// Unicode/Security/Covert/SurrogateReassembly.lean, mirrored from the
// verified Rust port's test vectors. An empty sub means clear
// (looksLikeByteStream gate off, or well-formed UTF-8); a non-empty sub is
// the first-violation tag.
func TestSurrogateReassemblyVectors(t *testing.T) {
	sub := func(input []uint32) string {
		subThreat, _, _ := surrogateReassemblyDetect(input)
		return subThreat
	}

	cases := []struct {
		name  string
		input []uint32
		want  string
	}{
		{"empty", []uint32{}, ""},
		{"ascii-hello", []uint32{0x48, 0x65, 0x6C, 0x6C, 0x6F}, ""},
		{"e-acute", []uint32{0xC3, 0xA9}, ""},
		{"han-zhong", []uint32{0xE4, 0xB8, 0xAD}, ""},
		{"emoji-grin", []uint32{0xF0, 0x9F, 0x98, 0x80}, ""},

		{"c0-80", []uint32{0xC0, 0x80}, "InvalidStartByte"},
		{"c0-af", []uint32{0xC0, 0xAF}, "InvalidStartByte"},
		{"fe", []uint32{0xFE}, "InvalidStartByte"},
		{"lone-continuation-80", []uint32{0x80}, "InvalidStartByte"},
		{"ff", []uint32{0xFF}, "InvalidStartByte"},

		{"overlong-slash-3byte", []uint32{0xE0, 0x80, 0xAF}, "Overlong"},
		{"overlong-slash-4byte", []uint32{0xF0, 0x80, 0x80, 0xAF}, "Overlong"},

		{"cesu8-surrogate-low", []uint32{0xED, 0xA0, 0x80}, "Cesu8"},
		{"cesu8-surrogate-high", []uint32{0xED, 0xAF, 0xBF}, "Cesu8"},

		{"truncated-2byte", []uint32{0xC3}, "Truncated"},
		{"truncated-4byte", []uint32{0xF0, 0x9F, 0x98}, "Truncated"},

		// The unit detect clamps values > 0xFF to 0xFF (mirroring the Lean
		// toBytes helper), which the strict decoder rejects as an invalid
		// start byte. The scan orchestrator gates these out (see below).
		{"non-byte-stream-emoji-codepoint", []uint32{0x1F600}, "InvalidStartByte"},
		{"non-byte-stream-mixed", []uint32{0x41, 0x100}, "InvalidStartByte"},
	}

	for _, tc := range cases {
		got := sub(tc.input)
		if got != tc.want {
			t.Errorf("%s: sub(%v) = %q, want %q", tc.name, tc.input, got, tc.want)
		}
	}
}

// TestSurrogateReassemblyByteStreamGate pins that the scan-orchestrator
// wrapper skips the family on non-byte-stream input (mirroring runAll),
// even though the unit detect clamps such input to a malformed byte.
func TestSurrogateReassemblyByteStreamGate(t *testing.T) {
	for _, input := range [][]uint32{{0x1F600}, {0x41, 0x100}} {
		if _, ok := surrogateReassemblyFinding(input); ok {
			t.Errorf("surrogateReassemblyFinding(%v) fired; want gated (no finding)", input)
		}
	}
	// A genuine byte-stream violation still produces a finding.
	if _, ok := surrogateReassemblyFinding([]uint32{0xC0, 0x80}); !ok {
		t.Errorf("surrogateReassemblyFinding([0xC0 0x80]) did not fire; want InvalidStartByte")
	}
}

File: ports/cpp/test/surrogate_reassembly_test.cpp
#include <doctest/doctest.h>

#include <cstdint>
#include <optional>
#include <span>
#include <string>
#include <vector>

#include "unicode_cpp/security/covert/surrogate_reassembly.hpp"

namespace sr = unicode_cpp::security::surrogate_reassembly;

// Ground truth: the `detect_*` spot-check theorems in
// `Unicode/Security/Covert/SurrogateReassembly.lean`, each proven by
// `decide`, and mirrored by the verified Rust port's test vectors.

static std::optional<std::string> sub(const std::vector<std::uint32_t> &input) {
  return sr::detect(std::span<const std::uint32_t>{input.data(), input.size()})
      .sub;
}

TEST_CASE("SurrogateReassembly — well-formed byte streams are clear") {
  CHECK(sub({}) == std::nullopt);
  CHECK(sub({0x48, 0x65, 0x6C, 0x6C, 0x6F}) == std::nullopt);  // "Hello"
  CHECK(sub({0xC3, 0xA9}) == std::nullopt);                    // é
  CHECK(sub({0xE4, 0xB8, 0xAD}) == std::nullopt);              // 中
  CHECK(sub({0xF0, 0x9F, 0x98, 0x80}) == std::nullopt);        // 😀
}

TEST_CASE("SurrogateReassembly — invalid start bytes") {
  CHECK(sub({0xC0, 0x80}) == std::optional<std::string>{"InvalidStartByte"});
  CHECK(sub({0xC0, 0xAF}) == std::optional<std::string>{"InvalidStartByte"});
  CHECK(sub({0xFE}) == std::optional<std::string>{"InvalidStartByte"});
  CHECK(sub({0x80}) == std::optional<std::string>{"InvalidStartByte"});
  CHECK(sub({0xFF}) == std::optional<std::string>{"InvalidStartByte"});
}

TEST_CASE("SurrogateReassembly — overlong encodings") {
  CHECK(sub({0xE0, 0x80, 0xAF}) == std::optional<std::string>{"Overlong"});
  CHECK(sub({0xF0, 0x80, 0x80, 0xAF}) ==
        std::optional<std::string>{"Overlong"});
}

TEST_CASE("SurrogateReassembly — CESU-8 surrogate codepoints") {
  CHECK(sub({0xED, 0xA0, 0x80}) == std::optional<std::string>{"Cesu8"});
  CHECK(sub({0xED, 0xAF, 0xBF}) == std::optional<std::string>{"Cesu8"});
}

TEST_CASE("SurrogateReassembly — truncated sequences") {
  CHECK(sub({0xC3}) == std::optional<std::string>{"Truncated"});
  CHECK(sub({0xF0, 0x9F, 0x98}) == std::optional<std::string>{"Truncated"});
}

TEST_CASE("SurrogateReassembly — non-byte-stream input is clear") {
  // Any entry >= 0x100 means this is a codepoint array, not a byte
  // stream, so the family does not apply.
  CHECK(sub({0x1F600}) == std::nullopt);
  CHECK(sub({0x41, 0x100}) == std::nullopt);
}
